diff -uNr cpio-2.4.2/configure.in cpio-2.4.2.new/configure.in
--- cpio-2.4.2/configure.in	Wed Dec 20 17:28:29 1995
+++ cpio-2.4.2.new/configure.in	Mon Aug 21 01:20:15 2000
@@ -62,6 +62,7 @@
 AC_FUNC_VPRINTF
 AC_FUNC_ALLOCA
 AC_HEADER_DIRENT
-AC_CHECK_LIB(nsl, gethostname, [LIBS="$LIBS -lnsl"])
+AC_CHECK_FUNC(gethostname, ,
+AC_CHECK_LIB(nsl, gethostname, [LIBS="$LIBS -lnsl"],) )
 AC_CHECK_LIB(socket, setsockopt, [LIBS="$LIBS -lsocket"])
 AC_OUTPUT(Makefile)
