--- cpio-2.4.2/copyin.c.stdout	Tue Aug 31 14:59:47 1999
+++ cpio-2.4.2/copyin.c	Tue Aug 31 15:00:52 1999
@@ -999,9 +999,9 @@
 	    }
 
 	  if (verbose_flag)
-	    fprintf (stderr, "%s\n", file_hdr.c_name);
+	    fprintf (stdout, "%s\n", file_hdr.c_name);
 	  if (dot_flag)
-	    fputc ('.', stderr);
+	    fputc ('.', stdout);
 	}
     }
 
--- cpio-2.4.2/copyout.c.stdout	Tue Aug 31 15:00:01 1999
+++ cpio-2.4.2/copyout.c	Tue Aug 31 15:02:27 1999
@@ -492,9 +492,9 @@
 	    }
 
 	  if (verbose_flag)
-	    fprintf (stderr, "%s\n", input_name.ds_string);
+	    fprintf (stdout, "%s\n", input_name.ds_string);
 	  if (dot_flag)
-	    fputc ('.', stderr);
+	    fputc ('.', stdout);
 	}
     }
 
